packages/@aws-cdk/aws-iam/lib/instance-profile.ts

```diff
@@ -0,0 +1,59 @@
+import { Construct } from '@aws-cdk/cdk';
+import { cloudformation } from './iam.generated';
+import { ServicePrincipal } from './policy-document';
+import { Role } from './role';
+
+export interface InstanceProfileProps {
+
+  /**
+   * The path associated with this instance profile. For information about IAM Instance Profiles, see
+   * Friendly Names and Paths in IAM User Guide.
+   */
+  path?: string;
+
+  /**
+   * The name of an existing IAM role to associate with this instance profile.
+   * Currently, you can assign a maximum of one role to an instance profile.
+   * @default Creates a default ec2.amazonaws.com Assume Role.
+   */
+  role?: Role;
+
+  /**
+   * The name of the instance profile that you want to create.
+   * This parameter allows (per its regex pattern) a string consisting of
+   * upper and lowercase alphanumeric characters with no spaces.
+   * You can also include any of the following characters: = , . @ -
+   */
+  instanceProfileName?: string;
+
+}
+
+/**
+ * IAM Instance Profile
+ *
+ * Defines an IAM Instance Profile that can be used with IAM roles for EC2 instances.
+ */
+export class InstanceProfile extends Construct {
+
+    public readonly roles = new Array<Role>();
+    public defaultRole = new Role(this, 'EC2Role', {
```

Comment:
I think we need to add a Role only if a role is not provided.
Also, I wasn't aware it is even possible to reference `this` from this context. Wow!

Anyway, I would write the ctor like this:

```ts
this.role = props.role || new Role(this, 'EC2Role', { assumedBy... });

new cloudformation.InstanceProfileResource(this, 'Resource', {
  instanceProfileName: props.instanceProfileName,
  roles: [ role.roleName ],
  path: props.path
});
```

Reply:
Cool.  I am new to typescript.  I am a seasoned java, JavaScript, etc developer.  So I naught be missing some things.